storey/targets.py

```diff
@@ -888,48 +891,100 @@ def _init(self):
-        if isinstance(value, datetime.datetime):
-            value = round(value.timestamp() * 1000)
-        elif isinstance(value, str):
-            value = f"'{value}'"
-        return str(value)
+    def _get_params_template(num_param: int) -> str:
+        return f"({','.join(num_param * ['?'])})"
 
-    async def _emit(self, batch, batch_key, batch_time, batch_events, last_event_time=None):
-        with StringIO() as b:
-            b.write("INSERT INTO ")
-            if self._table:
-                b.write(self._table)
-            else:  # table is dynamic
-                b.write(batch_key)
+    def _get_sql_template(self) -> str:
+        with StringIO() as sql:
+            sql.write("INSERT INTO ?")
             if self._supertable:
-                b.write(" USING ")
-                b.write(self._supertable)
-                b.write(" TAGS (")
-                for column_index in range(self._number_of_tags):
-                    value = batch[0].get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < self._number_of_tags - 1:
-                        b.write(",")
-                b.write(")")
-            b.write(" VALUES ")
-            for record in batch:
-                b.write("(")
-                for column_index in range(self._number_of_tags, len(self._columns)):
-                    value = record.get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < len(self._columns) - 1:
-                        b.write(",")
-                b.write(") ")
-            b.write(";")
-            insert_statement = b.getvalue()
-        self._connection.execute(insert_statement)
+                sql.write(f" USING {self._supertable} TAGS {self._get_params_template(self._number_of_tags)}")
+            sql.write(f" VALUES {self._get_params_template(self._number_of_values)};")
+            return sql.getvalue()
+
+    @classmethod
+    def _get_unsupported_error(cls, value) -> UnsupportedTDEngineTypeError:
+        return cls.UnsupportedTDEngineTypeError(f"Unsupported value type {type(value)} in {cls.__name__}")
+
+    @classmethod
+    def _value_to_tag(cls, value):
+        import taosws
+
+        if isinstance(value, bool):
+            return taosws.bool_to_tag(value)
+        if isinstance(value, int):
+            return taosws.int_to_tag(value)
+        if isinstance(value, float):
+            return taosws.float_to_tag(value)
+        if isinstance(value, str):
+            return taosws.varchar_to_tag(value)
+        raise cls._get_unsupported_error(value)
+
+    def _get_tags_from_event(self, event: dict) -> list:
+        tags = []
+        for tag_name in self._columns[: self._number_of_tags]:
+            tags.append(self._value_to_tag(event.get(tag_name)))
+        return tags
+
+    @classmethod
+    def _raw_value_to_column_fun(cls, value) -> Callable:
```

Comment:
```suggestion
    def _raw_value_to_column_func(cls, value) -> Callable:
```
(or function)

Applies elsewhere as well.